src/item.cpp

```diff
@@ -7330,3 +7346,17 @@ int item::get_min_str() const
         return type->min_str;
     }
 }
+
+requirement_data::alter_item_comp_vector item::get_uncraft_components() const
```

Comment:
I think a better alternative here is to return a `std::vector<item_comp>` rather than `requirement_data::alter_item_comp_vector`.  When `components.empty()` you have to extract the first element of each list of alternatives, rather than just return the recipe requirements verbatim.  That might seem like extra work, but actually it's extra code but less work because the current implementation has to copy the whole vector of vectors, which is probably more work for the program.

And overall it should be a code-size win, because the new code in the above implementation of `stacks_with` can be much simpler: just test `get_uncraft_components() != rhs.get_uncraft_components()`.  The other uses of `get_uncraft_components` should get slightly simpler too, but not such a big difference.

Reply:
Changed the auxiliary function to return a `std::vector<item_comp>` and modified calling code accordingly. I had to add equality operators to the `item_comp` struct for the vector comparison to work but it made the code more readable as you said.

About the `player:disassemble` thing, it only uses the recipe to check if the item can be disassembled. I left it that way because the new function only returns a list of materials without any information about the recipe so the recipe lookup is still needed. The actual disassembly is handled afterwards as an action and it uses the components in the item, the only thing that was wrong was the item list that was shown in the disassemble query, which is fixed already by using the new function to get the components list.